src/memory_region.rs

```diff
@@ -714,98 +404,153 @@ impl<'a> MemoryMapping<'a> {
-        match self {
-            MemoryMapping::Identity => ProgramResult::Ok(vm_addr),
-            MemoryMapping::Aligned(m) => m.map(access_type, vm_addr, len),
-            MemoryMapping::Unaligned(m) => m.map(access_type, vm_addr, len),
+        if let Some((_index, region)) = self.find_region(vm_addr) {
+            if let Some(host_addr) = region.vm_to_host(access_type, vm_addr, len) {
+                return ProgramResult::Ok(host_addr);
+            }
         }
+        let common = match &self {
+            MemoryMapping::Identity => return ProgramResult::Ok(vm_addr),
+            MemoryMapping::Aligned(m) => &m.common,
+            MemoryMapping::Unaligned(m) => &m.common,
+        };
+        generate_access_violation(common, access_type, vm_addr, len)
     }
 
-    /// Loads `size_of::<T>()` bytes from the given address.
+    /// Map virtual memory to host memory and potentially call the [AccessViolationHandler].
     ///
-    /// Works across memory region boundaries.
+    /// This requires the [MemoryMapping] to be mutable and
+    /// can cause previously translated addresses to become invalid.
     #[inline]
-    pub fn load<T: Pod + Into<u64>>(&self, vm_addr: u64) -> ProgramResult {
-        match self {
-            MemoryMapping::Identity => unsafe {
-                ProgramResult::Ok(ptr::read_unaligned(vm_addr as *const T).into())
-            },
-            MemoryMapping::Aligned(m) => m.load::<T>(vm_addr),
-            MemoryMapping::Unaligned(m) => m.load::<T>(vm_addr),
+    pub fn map_with_access_violation_handler(
+        &mut self,
+        access_type: AccessType,
+        vm_addr: u64,
+        len: u64,
+    ) -> ProgramResult {
+        let common = match &self {
+            MemoryMapping::Identity => return ProgramResult::Ok(vm_addr),
+            MemoryMapping::Aligned(m) => &m.common,
+            MemoryMapping::Unaligned(m) => &m.common,
+        };
+        if let Some((index, region)) = self.find_region(vm_addr) {
+            if let Some(host_addr) = region.vm_to_host(access_type, vm_addr, len) {
+                return ProgramResult::Ok(host_addr);
+            }
+            let mut region = (*region).clone();
+            let max_len = self
+                .get_regions()
+                .get(index.saturating_add(1))
+                .map_or(u64::MAX, |next_region| next_region.vm_addr)
+                .saturating_sub(region.vm_addr);
+            (common.access_violation_handler)(&mut region, max_len, access_type, vm_addr, len);
+            if let Err(err) = self.replace_region(index, region) {
+                return ProgramResult::Err(err);
+            }
+        }
+        // Give it a second try with the replaced region.
+        if let Some((_index, region)) = self.find_region(vm_addr) {
+            if let Some(host_addr) = region.vm_to_host(access_type, vm_addr, len) {
+                return ProgramResult::Ok(host_addr);
+            }
+        }
```

Comment:
Would it be the same if we called the access violation handler before the first check of `find_region`? I am wondering if we could have only one check.

Reply:
The `AccessViolationHandler` as the name implies, should only be called if an access violation occurred. That can only be found out by testing the region, thus the region needs to be found first.

Reply:
Let me formulate the question then. Since the vm_addr is not supposed to change, why do we have do find the region again? Isn't it going be the replaced one?

Reply:
One could potentially move the regions around. I know that we do currently not allow that in `replace_region`. This branch (of there being an access violation) is quite rare in the first place. So yes, we can potentially optimize it, but it could be dangerous if we ever allow moving vm addresses.

Reply:
> Thus, if you move a region upwards in virtual address space so that your access address now is located before it

That can only happen if the `access_violation_handler` changes the virtual address of a region, which it does not appear to be doing.